ibazel/command/notify_command.go

```diff
@@ -50,17 +52,18 @@ func (c *notifyCommand) Terminate() {
 		return
 	}
 
-	// Kill it with fire by sending SIGKILL to the process PID which should
-	// propagate down to any subprocesses in the PGID (Process Group ID). To
-	// send to the PGID, send the signal to the negative of the process PID.
-	// Normally I would do this by calling c.cmd.Process.Signal, but that
-	// only goes to the PID not the PGID.
-	c.pg.Kill()
+	c.pg.Kill(c.signum)
```

Comment:
1) I think we should probably rename the `Kill` command to something like `Signal` since you could send something like a `SIGHUP`, or `SIGUSR1` which are not kill commands.

2) Out of curiosity, what if we did something here this:

```
c.pg.Signal(SIGTERM)
var done chan bool
go func() {
  c.pg.Wait()
  done <- true
}
select {
case <- time.After(1*time.Second):
  c.pg.Signal(SIGKILL)
case <- done:
  // No action necessary since the job is gone on its own.
}
c.pg.Close()
```

I think that would handle the case where the process needs a small amount of time to shut down and also kill it if it really needs to die without adding a flag.

Can you tell me why my idea isn't sufficient, I think I might be missing something but I'm not sure?

Reply:
1. Makes sense. I'll rename. `Signal` sounds good to me.

2. I have played around with this idea as well. Technically it shouldn't cause any issues. Not even with a case of a Docker image, as Docker already got time to proxy the first signal to a container.

The downside of it would be to remove a control from a user. As an example, I'm personally planning to setup a development environment as close to a production environment as possible. That means using a Docker containers and graceful shutdown of servers. For that reason checking that graceful shutdown works as expected is important. If I would see operation stopping before it's completed, I would have to start investigating it. After a while it would turn out to be ibazel and I would be disappointed not being able to do anything about it.

This is a reason why I didn't implement it this way when ibazel detects SIGINT. SIGTERM and SIGHUP would actually result in a behavior like this. Instead of 1 second I used 10 seconds though. Now that I saw you using 1 second here in an example I started to think if 10 seconds was a bit too much?

Reply:
@vanils, thanks for the thoughtful response.

> The downside of it would be to remove a control from a user. As an example, I'm personally planning to setup a development environment as close to a production environment as possible. That means using a Docker containers and graceful shutdown of servers. For that reason checking that graceful shutdown works as expected is important. If I would see operation stopping before it's completed, I would have to start investigating it. After a while it would turn out to be ibazel and I would be disappointed not being able to do anything about it.

Hrm. These are very reasonable concerns. The only reason I'm pushing back on this is because right now the only control surfaces we have are basically to enable testing ibazel or to help develop ibazel in some other way. I don't believe you can even run -h to get a list of most of these flags.

* `--debounce` - Only intended to be used in tests to make them run faster
* `--log_to_file` - Used in testing to validate that the log output matches expectations
* `--nolive_reload` - Used to simplify some of the non-live-reload testcases (primarily controlled through bazel tags)
* `--bazel_path` -- Used in e2e testing to override the bazel path for when I was testing against many versions of bazel
* `--profile_dev` - Used to profile ibazel to find bottlenecks

There is really only one flag that does modifies control flow, `--run_output`. It exists because we can't defer trust to the repo. For example you clone a potentially malicious repo and you wanna see what happens when you do X, `ibazel` shouldn't prohibit you from doing things safely if you'd like to opt into it.

> This is a reason why I didn't implement it this way when ibazel detects SIGINT. SIGTERM and SIGHUP would actually result in a behavior like this. Instead of 1 second I used 10 seconds though. Now that I saw you using 1 second here in an example I started to think if 10 seconds was a bit too much?

What if we printed out an error message that said that the process wasn't killed within the X second window? Would that provide you the shutdown latency visibility you're seeking?

Reply:
Ahh. So maybe I missed the point of the your comment. So there would be 2 changes suggested.

- Use SIGTERM by default. Maybe removing the need for `--signal` flag in the first place?
- Use this timeout behavior with SIGINT, not just when detecting SIGTERM and SIGHUP?

Did I got it right? Your reasoning sounds good to me.

> What if we printed out an error message that said that the process wasn't killed within the X second window? Would that provide you the shutdown latency visibility you're seeking?

That sounds good. Maybe we could use higher timeout than 1 second? A user could always bail out faster by repeatedly hitting ctrl-c if they want. Maybe something like 5 seconds or even 10 seconds?

Reply:
> * Use SIGTERM by default. Maybe removing the need for --signal flag in the first place?

Anything to avoid flags 😉 

> * Use this timeout behavior with SIGINT, not just when detecting SIGTERM and SIGHUP?
> Did I got it right? Your reasoning sounds good to me.

Yep, I think that provides the needed task killing and also keeps the ergonomics of the current set up. Doing the right thing effectively silently (TM) the vast majority of the time is the goal here.

> That sounds good. Maybe we could use higher timeout than 1 second? A user could always bail out faster by repeatedly hitting ctrl-c if they want. Maybe something like 5 seconds or even 10 seconds?

Good point. The only issue with 10s is that it will mean we need to wait 10s in an e2e test to verify it and 10s is a long time for a test...

And now to completely contradict what I said just a few paragraphs ago... since that's controlling behavior that should be manipulated only in tests, would you be okay turning that into a flag but having a sane default of 10s?

Reply:
> And now to completely contradict what I said just a few paragraphs ago... since that's controlling behavior that should be manipulated only in tests, would you be okay turning that into a flag but having a sane default of 10s?

Sounds good! :) I'll mark this pr as a draft again and undraft it when all the changes have been done.